File: pkg/record/record.go
// Package record defines the DecisionRecord v1 data types and operations.
package record

import (
	"time"

	"github.com/google/uuid"
)

const SchemaVersion = "v1"

// DecisionRecord is the core evidence type for a single AI inference decision.
type DecisionRecord struct {
	SchemaVersion string        `json:"schema_version"`
	RequestID     uuid.UUID     `json:"request_id"`
	Timestamp     time.Time     `json:"timestamp"`
	Identity      Identity      `json:"identity"`
	AuthContext   *AuthContext  `json:"auth_context,omitempty"`
	Model         Model         `json:"model"`
	Parameters    Parameters    `json:"parameters"`
	PromptContext PromptContext `json:"prompt_context"`
	PolicyContext PolicyContext `json:"policy_context"`
	RAGContext    *RAGContext   `json:"rag_context,omitempty"`
	Output        Output        `json:"output"`
	Trace         Trace         `json:"trace"`
	Integrity     Integrity     `json:"integrity"`
}

type Identity struct {
	TenantID    string            `json:"tenant_id"`
	Subject     string            `json:"subject"`
	SubjectType string            `json:"subject_type,omitempty"`
	Claims      map[string]string `json:"claims,omitempty"`
}

// AuthContext captures server-populated authentication context to bind record
// identity to authenticated caller claims.
type AuthContext struct {
	Issuer        string `json:"issuer,omitempty"`
	Subject       string `json:"subject,omitempty"`
	TokenHash     string `json:"token_hash,omitempty"`
	Source        string `json:"source,omitempty"`
	Authenticated bool   `json:"authenticated,omitempty"`
}

type Model struct {
	Provider     string `json:"provider"`
	Name         string `json:"name"`
	Version      string `json:"version,omitempty"`
	Endpoint     string `json:"endpoint,omitempty"`
	DeploymentID string `json:"deployment_id,omitempty"`
}

type Parameters struct {
	Temperature      *float64 `json:"temperature,omitempty"`
	TopP             *float64 `json:"top_p,omitempty"`
	MaxTokens        *int     `json:"max_tokens,omitempty"`
	FrequencyPenalty *float64 `json:"frequency_penalty,omitempty"`
	PresencePenalty  *float64 `json:"presence_penalty,omitempty"`
	StopSequences    []string `json:"stop_sequences,omitempty"`
	Seed             *int     `json:"seed,omitempty"`
	ToolsEnabled     *bool    `json:"tools_enabled,omitempty"`
	ResponseFormat   string   `json:"response_format,omitempty"`
}

type PromptContext struct {
	SystemPromptHash       string `json:"system_prompt_hash,omitempty"`
	UserPromptHash         string `json:"user_prompt_hash"`
	UserPromptTemplateHash string `json:"user_prompt_template_hash,omitempty"`
	UserPromptTemplateID   string `json:"user_prompt_template_id,omitempty"`
	ToolSchemaHash         string `json:"tool_schema_hash,omitempty"`
	SafetyPromptHash       string `json:"safety_prompt_hash,omitempty"`
	MessageCount           int    `json:"message_count,omitempty"`
	TotalInputTokens       int    `json:"total_input_tokens,omitempty"`
}

type PolicyDecision string

const (
	PolicyAllow              PolicyDecision = "allow"
	PolicyDeny               PolicyDecision = "deny"
	PolicyAllowWithTransform PolicyDecision = "allow_with_transform"
	PolicyLogOnly            PolicyDecision = "log_only"
)

type PolicyContext struct {
	PolicyBundleID       string            `json:"policy_bundle_id,omitempty"`
	PolicyHash           string            `json:"policy_hash,omitempty"`
	PolicyDecision       PolicyDecision    `json:"policy_decision"`
	DecisionReasonCode   string            `json:"decision_reason_code,omitempty"`
	RuleIDs              []string          `json:"rule_ids,omitempty"`
	TransformsApplied    []TransformRecord `json:"transforms_applied,omitempty"`
	PolicyEngineVersion  string            `json:"policy_engine_version,omitempty"`
	EvaluationDurationMs float64           `json:"evaluation_duration_ms,omitempty"`
}

type TransformType string

const (
	TransformRedactPII TransformType = "redact_pii"
	TransformRedactPHI TransformType = "redact_phi"
	TransformMask      TransformType = "mask"
	TransformFilter    TransformType = "filter"
	TransformCustom    TransformType = "custom"
)

type TransformTarget string

const (
	TransformTargetInput  TransformTarget = "input"
	TransformTargetOutput TransformTarget = "output"
	TransformTargetBoth   TransformTarget = "both"
)

type TransformRecord struct {
	Type    TransformType   `json:"type"`
	Target  TransformTarget `json:"target"`
	Details string          `json:"details,omitempty"`
}

type RAGContext struct {
	ConnectorIDs            []string              `json:"connector_ids,omitempty"`
	DocumentIDs             []string              `json:"document_ids,omitempty"`
	ChunkHashes             []string              `json:"chunk_hashes,omitempty"`
	CitationHashes          []string              `json:"citation_hashes,omitempty"`
	RetrievalPolicyDecision string                `json:"retrieval_policy_decision,omitempty"`
	PromptInjectionCheck    *PromptInjectionCheck `json:"prompt_injection_check,omitempty"`
}

type PromptInjectionCheck struct {
	Performed       bool   `json:"performed"`
	Result          string `json:"result"`
	DetectorVersion string `json:"detector_version,omitempty"`
}

type OutputMode string

const (
	OutputModeHashOnly  OutputMode = "hash_only"
	OutputModeEncrypted OutputMode = "encrypted"
	OutputModePlaintext OutputMode = "plaintext"
)

type Output struct {
	OutputHash          string     `json:"output_hash"`
	Mode                OutputMode `json:"mode"`
	OutputEncrypted     string     `json:"output_encrypted,omitempty"`
	OutputEncryptedRef  string     `json:"output_encrypted_ref,omitempty"`
	OutputEncryptedHash string     `json:"output_encrypted_hash,omitempty"`
	OutputPlaintext     string     `json:"output_plaintext,omitempty"`
	OutputTokens        int        `json:"output_tokens,omitempty"`
	FinishReason        string     `json:"finish_reason,omitempty"`
	LatencyMs           float64    `json:"latency_ms,omitempty"`
}

type Trace struct {
	OtelTraceID     string `json:"otel_trace_id,omitempty"`
	OtelSpanID      string `json:"otel_span_id,omitempty"`
	ParentRequestID string `json:"parent_request_id,omitempty"`
	SessionID       string `json:"session_id,omitempty"`
}

type Integrity struct {
	SequenceNumber     int64           `json:"sequence_number,omitempty"`
	RecordHash         string          `json:"record_hash"`
	PreviousRecordHash string          `json:"previous_record_hash,omitempty"`
	MerkleRoot         string          `json:"merkle_root,omitempty"`
	MerkleTreeSize     int64           `json:"merkle_tree_size,omitempty"`
	InclusionProofRef  string          `json:"inclusion_proof_ref,omitempty"`
	InclusionProof     *InclusionProof `json:"inclusion_proof,omitempty"`
}

type InclusionProof struct {
	LeafIndex int64    `json:"leaf_index"`
	Hashes    []string `json:"hashes"`
}

// Receipt is returned to the client after a record is appended to the ledger.
type Receipt struct {
	RequestID         uuid.UUID       `json:"request_id"`
	SequenceNumber    int64           `json:"sequence_number"`
	RecordHash        string          `json:"record_hash"`
	MerkleRoot        string          `json:"merkle_root"`
	MerkleTreeSize    int64           `json:"merkle_tree_size"`
	InclusionProofRef string          `json:"inclusion_proof_ref,omitempty"`
	InclusionProof    *InclusionProof `json:"inclusion_proof,omitempty"`
	Timestamp         time.Time       `json:"timestamp"`
}

// New creates a new DecisionRecord with the schema version set and a generated request ID.
func New() *DecisionRecord {
	return &DecisionRecord{
		SchemaVersion: SchemaVersion,
		RequestID:     uuid.New(),
		Timestamp:     time.Now().UTC(),
	}
}

File: sdk/python/vaol/record.py
"""DecisionRecord builder for constructing VAOL evidence records."""

from __future__ import annotations

import hashlib
import json
import uuid
from datetime import UTC, datetime
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field


class OutputMode(StrEnum):
    HASH_ONLY = "hash_only"
    ENCRYPTED = "encrypted"
    PLAINTEXT = "plaintext"


class PolicyDecision(StrEnum):
    ALLOW = "allow"
    DENY = "deny"
    ALLOW_WITH_TRANSFORM = "allow_with_transform"
    LOG_ONLY = "log_only"


class Identity(BaseModel):
    tenant_id: str
    subject: str
    subject_type: str = "user"
    claims: dict[str, str] = Field(default_factory=dict)


class AuthContext(BaseModel):
    issuer: str = ""
    subject: str = ""
    token_hash: str = ""
    source: str = ""
    authenticated: bool = False


class ModelInfo(BaseModel):
    provider: str
    name: str
    version: str = ""
    endpoint: str = ""
    deployment_id: str = ""


class Parameters(BaseModel):
    temperature: float | None = None
    top_p: float | None = None
    max_tokens: int | None = None
    frequency_penalty: float | None = None
    presence_penalty: float | None = None
    stop_sequences: list[str] = Field(default_factory=list)
    seed: int | None = None
    tools_enabled: bool | None = None
    response_format: str = ""


class PromptContext(BaseModel):
    system_prompt_hash: str = ""
    user_prompt_hash: str
    user_prompt_template_hash: str = ""
    user_prompt_template_id: str = ""
    tool_schema_hash: str = ""
    safety_prompt_hash: str = ""
    message_count: int = 0
    total_input_tokens: int = 0


class PolicyContext(BaseModel):
    policy_bundle_id: str = ""
    policy_hash: str = ""
    policy_decision: PolicyDecision = PolicyDecision.LOG_ONLY
    decision_reason_code: str = ""
    rule_ids: list[str] = Field(default_factory=list)
    transforms_applied: list[dict[str, str]] = Field(default_factory=list)
    policy_engine_version: str = ""
    evaluation_duration_ms: float = 0


class Output(BaseModel):
    output_hash: str
    mode: OutputMode = OutputMode.HASH_ONLY
    output_encrypted: str = ""
    output_encrypted_ref: str = ""
    output_encrypted_hash: str = ""
    output_plaintext: str = ""
    output_tokens: int = 0
    finish_reason: str = ""
    latency_ms: float = 0


class Trace(BaseModel):
    otel_trace_id: str = ""
    otel_span_id: str = ""
    parent_request_id: str = ""
    session_id: str = ""


class Integrity(BaseModel):
    record_hash: str = ""
    sequence_number: int = 0
    previous_record_hash: str = ""
    merkle_root: str = ""
    merkle_tree_size: int = 0
    inclusion_proof_ref: str = ""


class DecisionRecord(BaseModel):
    """A VAOL DecisionRecord v1."""

    schema_version: str = "v1"
    request_id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    timestamp: str = Field(default_factory=lambda: datetime.now(UTC).isoformat())
    identity: Identity
    auth_context: AuthContext | None = None
    model: ModelInfo
    parameters: Parameters = Field(default_factory=Parameters)
    prompt_context: PromptContext
    policy_context: PolicyContext = Field(default_factory=PolicyContext)
    rag_context: dict[str, Any] | None = None
    output: Output
    trace: Trace = Field(default_factory=Trace)
    integrity: Integrity = Field(default_factory=Integrity)

    def to_dict(self) -> dict[str, Any]:
        """Serialize to a dictionary, excluding None values."""
        return self.model_dump(exclude_none=True)

    def to_json(self) -> str:
        """Serialize to JSON string."""
        return self.model_dump_json(exclude_none=True)


def sha256_hash(data: str | bytes) -> str:
    """Compute a SHA-256 hash with the 'sha256:' prefix."""
    if isinstance(data, str):
        data = data.encode("utf-8")
    digest = hashlib.sha256(data).hexdigest()
    return f"sha256:{digest}"


def hash_messages(messages: list[dict[str, Any]]) -> str:
    """Hash a list of chat messages (OpenAI format) deterministically."""
    canonical = json.dumps(messages, sort_keys=True, separators=(",", ":"))
    return sha256_hash(canonical)
